zipline/utils/events.py

```diff
@@ -522,85 +501,74 @@ def get_last_trading_day_of_week(dt, env):
+            return env.previous_trading_day(prev)
 
     date_func = get_last_trading_day_of_week
 
 
-class NthTradingDayOfMonth(StatelessRule):
-    """
-    A rule that triggers on the nth trading day of the month.
-    This is zero-indexed, n=0 is the first trading day of the month.
-    """
+class TradingDayOfMonthRule(six.with_metaclass(ABCMeta, StatelessRule)):
     def __init__(self, n=0):
         if not 0 <= n < MAX_MONTH_RANGE:
             raise _out_of_range_error(MAX_MONTH_RANGE)
-        self.td_delta = n
         self.month = None
-        self.day = None
+        self.date = None
 
     def should_trigger(self, dt, env):
-        return self.get_nth_trading_day_of_month(dt, env) == dt.date()
+        return self.get_trigger_day_of_month(dt, env) == env.normalize_date(dt)
+
+    @abstractmethod
+    def date_func(self, dt, env):
+        raise NotImplementedError
 
-    def get_nth_trading_day_of_month(self, dt, env):
+    def get_trigger_day_of_month(self, dt, env):
         if self.month == dt.month:
             # We already computed the day for this month.
-            return self.day
+            return self.date
 
-        if not self.td_delta:
-            self.day = self.get_first_trading_day_of_month(dt, env)
-        else:
-            self.day = env.add_trading_days(
-                self.td_delta,
-                self.get_first_trading_day_of_month(dt, env),
-            ).date()
+        self.date = self.date_func(dt, env)
+        if self.td_delta and self.date:
+            try:
+                self.date = env.add_trading_days(self.td_delta, self.date)
+            except NoFurtherDataError:
+                self.date = None
+
+        return self.date
 
-        return self.day
+
+class NthTradingDayOfMonth(TradingDayOfMonthRule):
+    """
+    A rule that triggers on the nth trading day of the month.
+    This is zero-indexed, n=0 is the first trading day of the month.
+    """
+    def __init__(self, n=0):
+        super(NthTradingDayOfMonth, self).__init__(n)
+        self.td_delta = n
 
     def get_first_trading_day_of_month(self, dt, env):
         self.month = dt.month
 
         dt = dt.replace(day=1)
-        self.first_day = (dt if env.is_trading_day(dt)
-                          else env.next_trading_day(dt)).date()
-        return self.first_day
+        first_day = (dt if env.is_trading_day(dt)
+                     else env.next_trading_day(dt))
+        return first_day
 
+    date_func = get_first_trading_day_of_month
 
-class NDaysBeforeLastTradingDayOfMonth(StatelessRule):
+
+class NDaysBeforeLastTradingDayOfMonth(TradingDayOfMonthRule):
     """
     A rule that triggers n days before the last trading day of the month.
     """
     def __init__(self, n=0):
-        if not 0 <= n < MAX_MONTH_RANGE:
-            raise _out_of_range_error(MAX_MONTH_RANGE)
+        super(NDaysBeforeLastTradingDayOfMonth, self).__init__(n)
         self.td_delta = -n
```

Comment:
Probably more consistent with the other attrs to have the base class do the assignment.  Could include a flag for whether to invert it.  What do you think?